### DISCUSSION TITLE
Queue

### DISCUSSION THREAD

**Discussion by daiyam** [basharovV/musicat#55] [General]
Hello,

I don't quite understand the queue behaviour...

The `Custom queue` tab/button clears the queue to create a new playlist with the input underneath, I assume.
The `Same as library` tab/button replace the current queue with all the songs.

I believe they should be options to a right-click menu, like:
- Create a new queue/playlist
- Save as playlist
- Same as library

Also the types `PlaylistType` and `QueueMode` have almost the same usage (`PlaylistType` is used to display the queue's mode...)
To have 2 playlists `playlist` and `albumPlaylist` is quite confusing.

I happy to help you to streamline it :wink:

**Comment by basharovV** [comment]
Yes, there's a few weird behaviours in there, which I admit are a bit inconsistent. Let's make it better :) 

Just to explain the current modes: 
- **library mode**. Clicking on "same as library" resets the queue to the same tracks as in the library order. Shuffling then unshuffling will always return to the original library order. 
     - _weirdness_ ❗: You can still add, re-order, and remove tracks in this view. 
- **custom queue mode**. The idea behind the Custom queue mode was to use the queue view as a sort of preliminary-playlist - you can add, re-order and delete tracks in the queue, but playing another track in the library won't affect this queue_. And if you want it as a playlist you can easily create a playlist with those tracks. 
   -  _weirdness_ ❗: Queue is not persisted when switching back after library mode.  
- **album mode**. Resets the queue to the album tracks only. 
    - _weirdness_ ❗:  Again, you can edit the order, but it goes against the principle of having a separate custom queue. 
    - _bug_ ❗: Overrides Custom queue. Ideally it would just change the mode to "library mode" and leave custom queue untouched, in case we want to come back to it. 

Conclusion - it's confusing as hell! I think you're right, maybe we can simplify it to a **single editable queue** that: 
- **resets whenever you choose music manually**. For example: 
   - double-click on track in library -> queue mirrors library. 
   - double click on album -> just want to listen to album start to finish, so queue becomes album. 
   - double-click track in playlist -> queue reset to playlist
- if you don't want the queue to be reset - then drag and drop tracks into it, or right-click track-> add to queue.
- make this configurable in settings? "reset queue on manual track selection" [ x ] . When turned off, playing a new track/album would add it to the top of the existing queue. 

Queue options: 
1) Clear the queue
2) Reset queue to library order (or just double click on a track in the library to do the same). 
3) Save as playlist

**Reply by daiyam** [reply]
> * **resets whenever you choose music manually**. For example:
>   
>   * double-click on track in library -> queue mirrors library.
>   * double click on album -> just want to listen to album start to finish, so queue becomes album.
>   * double-click track in playlist -> queue reset to playlist
> * if you don't want the queue to be reset - then drag and drop tracks into it, or right-click track-> add to queue.

I validate. It's simpler.

> * make this configurable in settings? "reset queue on manual track selection" [ x ] . When turned off, playing a new track/album would add it to the top of the existing queue.

Hum, the option can be confusing... Why on the top and not at the end?
I would avoid it.
 
> Queue options:
> 
> 1. Clear the queue
> 2. Reset queue to library order (or just double click on a track in the library to do the same).
> 3. Save as playlist

Ok

**Comment by daiyam** [comment]
> double-click on track in library -> queue mirrors library.

Currently, the queue starts from the clicked track but:
- it didn't display preceding tracks
- `playPrevious` is still available for the preceding track

Is it ok to change to:
- preceding tracks can be viewed by scrolling up

Should the whole library become the queue? Should it be limited? In size?

**Reply by basharovV** [reply]
I agree let's change it to show the whole list, auto scroll to the current track on open and allow scrolling up, that's how other music apps do it too (eg. YouTube music). Usually they also slightly fade out the above tracks.

**Comment by daiyam** [comment]
Last question of the year ;)

I found out that operations (add/remove songs) on the shuffled queue aren't replicated on the original queue. So when unshuffled, all the operations will be lost. Is it the expected behaviour?

**Reply by basharovV** [reply]
Happy new year! Good point, I would expect the queue operations to be mirrored across both queues. That's a bit more tricky though, especially with multiple-selection, but you just find the indexes of the songs affected in the other array. 

Alternatively, we get rid rid of shufflePlaylist and only use the one `playlist`, shuffle it in place, and when un-shuffling we simply re-sort?

**Comment by daiyam** [comment]
done in #70
